lib/icinga/clusterevents-check.cpp

```diff
@@ -98,8 +98,84 @@ void ClusterEvents::ExecuteCheckFromQueue(const MessageOrigin::Ptr& origin, cons
 	}
 
 	if (params->Contains("source")) {
-		Log(LogCritical, "ApiListener", "Not implemented.");
-		return;
+		String uuid = params->Get("source");
+
+		Host::Ptr host = Host::GetByName(params->Get("host"));
+		if (!host) {
+			Log(LogCritical, "ApiListener", "Host not found.");
+			return;
+		}
+
+		Checkable::Ptr checkable;
+		if (params->Contains("service"))
+			checkable = host->GetServiceByShortName(params->Get("service"));
+		else
+			checkable = host;
+
+		if (!checkable) {
+			Log(LogCritical, "ApiListener", "Checkable not found.");
+			return;
+		}
+
+		ObjectLock oLock (checkable);
+
+		if (origin->FromZone && !origin->FromZone->CanAccessObject(checkable)) {
+			Log(LogNotice, "ApiListener")
+				<< "Discarding 'ExecuteCheckFromQueue' event for checkable '" << checkable->GetName()
+				<< "' from '" << origin->FromClient->GetIdentity() << "': Unauthorized access.";
+			return;
+		}
+
+		/* Check deadline */
+		double deadline = params->Get("deadline");
+		if (Utility::GetTime() > deadline) {
+			Log(LogNotice, "ApiListener")
+				<< "Discarding 'ExecuteCheckFromQueue' event for checkable '" << checkable->GetName()
+				<< "' from '" << origin->FromClient->GetIdentity() << "': Deadline has expired.";
+			return;
+		}
+
+		Checkable::ExecuteCommandProcessFinishedHandler = [listener, sourceEndpoint, origin, params] (const Checkable::Ptr& checkable, const CheckResult::Ptr& cr, const Value& commandLine, const ProcessResult& pr) -> void {
```

Comment:
You need to make sure this gets reset. Outside of the if make a local bool variable (default false), defer a lambda capturing this variable by ref. which resets this one if the variable is true and after setting this one set the variable to true.